### DISCUSSION TITLE
[question] How to iterate map variable and get key, value pair?

### DISCUSSION THREAD

**Discussion by benet96** [bpftrace/bpftrace#2050] [Q&A] [answered]
I only find function `print` to print out map variable.

It said `@` will expand to one or more key and value. In C language, a comma expression will return the last variable's value, but `(@)` is illegal. So how could I iterate map and get key and value pair?

**Comment by fbs** [comment, marked as answer]
Hi, what are you trying to solve?

Afaik there is no wait to iterate maps in BPF, you can in user space (which is how we print maps) but user space behaviour is not programmable

**Reply by danobi** [reply]
Actually, you can in BPF now: https://github.com/torvalds/linux/blob/a379fbbcb88bcf43d886977c6fb91fb43f330b84/include/uapi/linux/bpf.h#L4716-L4743

It's possible that bpftrace could support it but we'd need a strong reason to do so. We might also need to implement user defined functions to support it.

**Reply by benet96** [reply]
> Actually, you can in BPF now: https://github.com/torvalds/linux/blob/a379fbbcb88bcf43d886977c6fb91fb43f330b84/include/uapi/linux/bpf.h#L4716-L4743
> 
> It's possible that bpftrace could support it but we'd need a strong reason to do so. We might also need to implement user defined functions to support it.

Thank you, I got it.

**Reply by benet96** [reply]
So far, I can not iterate any map in 'bpftrace' language, but 'C language' APIs can do it.
